### PR TITLE
Remove Lodash from client/login

### PR DESCRIPTION
**Pull request by jsnajdr**
This PR removes Lodash usage from the `client/login` directory. Also does two drive-by improvements in `RequestLoginEmailForm`: migrate a legacy lifecycle, and use a modern ref for reffing the input field.

**How to test:**
I tested mainly the `/log-in/link` page. Entered an invalid username so that error is triggered, and verified that the username field is refocused after failed request:

Screenshot 2022-01-20 at 10 53 54

Testing this also proves that the `defer` call wasn't needed: we can call `el.focus()` directly in the `componentDidUpdate` method.

### CODE REVIEW

File: client/login/wp-login/login-links.jsx [hunk 1]
```diff
@@ -350,9 +349,8 @@ export default connect(
 		isLoggedIn: Boolean( getCurrentUserId( state ) ),
 		oauth2Client: getCurrentOAuth2Client( state ),
 		query: getCurrentQueryArguments( state ),
-		isJetpackWooCommerceFlow:
-			'woocommerce-onboarding' === get( getCurrentQueryArguments( state ), 'from' ),
-		wccomFrom: get( getCurrentQueryArguments( state ), 'wccom-from' ),
+		isJetpackWooCommerceFlow: 'woocommerce-onboarding' === getCurrentQueryArguments( state ).from,
```
Comment (tyxla) [inline client/login/wp-login/login-links.jsx:352]: `getCurrentQueryArguments( state )` selector could return `null`, and accessing its `.from` would trigger an error. Should we use optional changing to prevent that?
```suggestion
		isJetpackWooCommerceFlow: 'woocommerce-onboarding' === getCurrentQueryArguments( state )?.from,
```

File: client/login/wp-login/login-links.jsx [hunk 2]
```diff
@@ -350,9 +349,8 @@ export default connect(
...
+		wccomFrom: getCurrentQueryArguments( state )[ 'wccom-from' ],
```
Comment (tyxla) [inline client/login/wp-login/login-links.jsx:353]: Just like above.

```suggestion
		wccomFrom: getCurrentQueryArguments( state )?.[ 'wccom-from' ],
```

File: client/login/ssr.js [hunk 3]
```diff
@@ -33,7 +29,7 @@ export function setShouldServerSideRenderLogin( context, next ) {
  * @returns {boolean} If the value of &redirect_to= on the log-in page is compatible with SSR
  */
 function isRedirectToValidForSsr( redirectToQueryValue ) {
-	if ( 'undefined' === typeof redirectToQueryValue ) {
+	if ( redirectToQueryValue === undefined ) {
```
Comment (tyxla) [inline client/login/ssr.js:32]: Won't that trigger a `Uncaught ReferenceError: redirectToQueryValue is not defined` error?

**Review by tyxla — COMMENTED** [review]
Nice work! Looks good overall, left some comments.

File: client/login/wp-login/login-links.jsx (hunk 1 above)
Comment (jsnajdr) [reply client/login/wp-login/login-links.jsx:352]: The selector will return either `false` or an object, never `null`. And field access on `false` is possible.

We already use direct `getCurrentQueryArguments( state ).email_address` at several places in Login and it's safe.

I can add `?.from` but we should rather migrate towards state where it's not needed. After all, the app is always on _some_ route, so there's no reason why the current path or query should be `null`.

File: client/login/ssr.js (hunk 3 above)
Comment (jsnajdr) [reply client/login/ssr.js:32]: No, the `redirectToQueryValue` identifier is always defined because it's the function's parameter. The `typeof` check is only useful for globals like `window` that don't need to exist at all.

File: client/login/wp-login/login-links.jsx (hunk 1 above)
Comment (tyxla) [reply client/login/wp-login/login-links.jsx:352]: > The selector will return either false or an object, never null. 

If that's the case, should we update the selector? It currently defaults to `null`: 

https://github.com/Automattic/wp-calypso/blob/3eb87c8860b8f24db1d592040af747688dc575e7/client/state/selectors/get-current-query-arguments.js#L11

> We already use direct getCurrentQueryArguments( state ).email_address at several places in Login and it's safe.

That's fair 👍 

> I can add ?.from but we should rather migrate towards state where it's not needed. After all, the app is always on some route, so there's no reason why the current path or query should be null.

Agreed. I'm not sure why the selector defaults to `null` TBH 🤷

File: client/login/ssr.js (hunk 3 above)
Comment (tyxla) [reply client/login/ssr.js:32]: Then why don't we remove the check? I just don't understand why we replace the `typeof x === 'undefined'` check with a `=== undefined`. How is the latter better?

Comment (jsnajdr) [reply client/login/ssr.js:32]: We need the check: if `redirectToQueryValue` is `undefined`, the result is immediately `true` because the URL doesn't have `redirectTo` param to validate. Only when it has a "real" value do we proceed to validate it.

The `=== undefined` check is simpler and sufficient. The `typeof x` check is needed only for special purposes, and in our codebase we overuse it a lot.

Comment (tyxla) [reply client/login/ssr.js:32]: In my head `typeof x === 'undefined` is much safer and I always prefer it to `=== undefined`. The only reason why one would use the latter would be because it's a bit shorter, but always prefer the safer than shorter TBH. Unless you have a better reason than brevity?

Comment (tyxla) [reply client/login/ssr.js:32]: Noting that as long `redirectToQueryValue` is always declared, this convo is effectively nitpicking 😉

File: client/login/magic-login/request-login-email-form.jsx [hunk 4]
```diff
@@ -49,10 +48,11 @@ class RequestLoginEmailForm extends Component {
...
 
-	// @TODO: Please update https://github.com/Automattic/wp-calypso/issues/58453 if you are refactoring away from UNSAFE_* lifecycle methods!
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		if ( ! this.props.requestError && nextProps.requestError ) {
-			defer( () => this.usernameOrEmail && this.usernameOrEmail.focus() );
+	usernameOrEmailRef = createRef();
+
+	componentDidUpdate( prevProps ) {
+		if ( ! prevProps.requestError && this.props.requestError ) {
+			this.usernameOrEmailRef.current?.focus();
```
Comment (tyxla) [inline client/login/magic-login/request-login-email-form.jsx:55]: This is similar behavior as with the`autoFocus` prop. Generally, we should avoid that for accessibility reasons, but it would be a job for another PR, as this component was already doing it.

**Review by tyxla — APPROVED** [review]
This works well and the only comments I had were non-blocking nits. Nice work!

File: client/login/magic-login/request-login-email-form.jsx (hunk 4 above)
Comment (jsnajdr) [reply client/login/magic-login/request-login-email-form.jsx:55]: So we should remove all the auto-focusing and re-focusing code altogether from this UI? If we agree it's a good idea, I'd do in this PR as another commit, it's a welcome simplification.

File: client/login/ssr.js (hunk 3 above)
Comment (jsnajdr) [reply client/login/ssr.js:32]: > In my head `typeof x === 'undefined` is much safer and I always prefer it to `=== undefined`.

I think you should start using `=== undefined` 🙂 The only safety added by `typeof` is that the `typeof blah_blah` expression is valid also on identifiers that are not declared at all. But most of the time you can be 100% sure that the identifier is declared. There is a parameter declaration or variable declaration nearby.

File: client/login/magic-login/request-login-email-form.jsx (hunk 4 above)
Comment (tyxla) [reply client/login/magic-login/request-login-email-form.jsx:55]: Yeah, I think removing the auto-focusing would actually be an improvement 👍
